feat: Add changelog notifier integration

rolysr opened the pull request:
This PR introduces changes on the master side to allow registration of a new type of application that receives custom changelog messages for metadata operations. The implementation enables external applications to subscribe to metadata events and receive messages with corresponding file paths resolved by inode, improving integration and observability for metadata changes.

rolysr committed:
feat: Add changelog notifier integration

This commit introduces changes on the master side to allow
registration of a new type of application that receives custom
changelog messages for metadata operations. The implementation
enables external applications to subscribe to metadata events
and receive messages with corresponding file paths resolved by
inode, improving integration and observability for metadata
changes.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

antuan96314 committed:
feat: Add support for inotifier services

 - Add table in server session for cgi.
 - Add list-inotifiers option to `saunafs-admin` command:
	i.e. `saunafs-admin list-inotifiers sfsmaster 9421`

Co-authored-by: Rolando Sánchez Ramos <rolysr@leil.io>

Signed-off-by: Antuan <antuan@leil.io>

uristdwarf commented on src/data/sfsmaster.cfg.in:
```diff
@@ -93,6 +93,10 @@
 
+## Port to listen on for notifier connections.
+## (Default: 9418)
+# MATONT_LISTEN_PORT = 9418
```
Remember to document this as experimental!

rolysr replied:
Done.

rolysr committed:
feat: Add path by inode request to notifiers

This commit adds a new type of request to the master-notifier
communication, allowing the master to send the type and full
path for a specific inode requested by a notifier application.
This change provides additional utility for obtaining more
detailed analysis from changelog texts and these requests.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

rolysr committed:
chore: Bump version to 5.4.0

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	int sock;
+	int32_t pdescpos;
+	uint32_t lastread, lastwrite;
+	uint8_t hdrbuff[kPacketHeaderSize];
```
Use std::array instead

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	uint32_t lastread, lastwrite;
+	uint8_t hdrbuff[kPacketHeaderSize];
+	PacketStruct inputpacket;
+	std::list<PacketStruct> outputPackets;
```
Try not to use std::list, use std::deque/std::queue or std::vector for better cache performance

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	      servip(0),
+	      config() {}
+
+	~MatontservEntry() = default;
```
Is this needed?

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	~MatontservEntry() = default;
+};
+
+static std::list<MatontservEntry> matontservList;
```
See my other comment regarding std::list

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+static std::string ListenHost;
+static std::string ListenPort;
+
+uint8_t *matontserv_addpacket(MatontservEntry *eptr, uint32_t type, uint32_t size) {
```
If we are going to create a new service with C++ entirely, why not also change the style to our own as well (e.g camelCase?)

rolysr replied:
Considered. For now I will keep this change as optional for the ease of the review and keep minimal correspondence with previous refactored code for other master services.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+		data = matontserv_addpacket(&entry, MATONT_METACHANGES_LOG, 9 + message.size());
+		put8bit(&data, 0xFF);
+		put64bit(&data, version);
+		memcpy(data, message.c_str(), message.size());
```
Why not try appending to the buffer vector directly (e.g buffer.insert), instead of using raw pointers? At the very least that avoids any buffer overwrites (though make sure you have enough memory in the buffer, otherwise you risk allocating a lot and it could cause memory leaks.

rolysr replied:
For now I prefer to keep this logic simpler as it is for first iteration.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+
+// from config
+static std::string ListenHost;
+static std::string ListenPort;
```
I'd consider creating a single struct that is global for these elements, rather than having multiple globals.

rolysr replied:
Considered.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	if (oldListenHost == ListenHost && oldListenPort == ListenPort) {
+		safs_pretty_syslog(LOG_NOTICE,
+		                   "master <-> notifiers module: socket address hasn't changed (%s:%s)",
+		                   ListenHost.c_str(), ListenPort.c_str());
```
Use the safs::log functions instead

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	matontservList.clear();
+}
+
+void matontserv_desc(std::vector<pollfd> &pdesc) {
+	if (!gExiting) {
+		pdesc.push_back({lsock, POLLIN, 0});
+		lsockpdescpos = pdesc.size() - 1;
+	} else {
+		lsockpdescpos = -1;
+	}
+	for (auto &entry : matontservList) {
+		pdesc.push_back({entry.sock, POLLIN, 0});
+		entry.pdescpos = pdesc.size() - 1;
+		if (!entry.outputPackets.empty()) { pdesc.back().events |= POLLOUT; }
+	}
+}
```
Consider renaming the variables here to better describe them

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	pt[1] = (ip >> 16) & 0xFF;
+	pt[2] = (ip >> 8) & 0xFF;
+	pt[3] = ip & 0xFF;
+	return std::to_string(pt[0]) + "." + std::to_string(pt[1]) + "." + std::to_string(pt[2]) + "." +
```
Feel like fmt::format would work better here

rolysr replied:
Fixed by reusing the already defined function `ipToString`

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	}
+	safs_pretty_syslog(LOG_WARNING, "no notifiers connected !!!");
+}
+
```
[IMPORTANT] I think it's perfectly normal not to have inotifiers connected, so you don't need this

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+void matontserv_become_master() {
+	eventloop_timeregister(TIMEMODE_SKIP_LATE, 3600, 0, matontserv_status);
+	return;
+}
```
[IMPORTANT] Also not needed

rolysr replied:
Done.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+}
+
+void matontserv_register(MatontservEntry *eptr, const uint8_t *data, uint32_t length) {
+	if (eptr->version > 0) {
```
Is version checking really needed?

uristdwarf replied:
Maybe in the future we could do without it.

rolysr replied:
Yes, it is needed because is the way we have to validate an already established connection. Note that this version field is updated only during register of a new notifier initiating a connection, so this manage edge case of same notifier sending a register when the master already has its connection registered. On the other hand this field is also very important on `matontserv_get_path_type_inode`  where it is use for determine if that received request came from an already registered notifier and not from another app or process initiation a connection for responding to that request only, which is an optimization for this case.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+			safs_pretty_syslog(
+			    LOG_NOTICE, "master <-> notifiers module: got unknown message (type:%" PRIu32 ")",
+			    type);
+			eptr->mode = NotifierConnectionMode::KILL;
```
We should start sending ANTOAN_UNKNOW_COMMAND and handling that on the other end, rather than killing the connection

rolysr replied:
Considered. This is a good and more robust consideration for next iteration maybe.

uristdwarf commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+		}
+		if (i < 0) {
+			if (errno != EAGAIN) {
+				safs_silent_errlog(LOG_NOTICE, "read from NT(%s) error", eptr->servstrip.c_str());
```
[IMPORTANT] Log the actual errno string

rolysr replied:
Ok I will use for this: `safs::log_info_with_error_code`

rolysr replied:
Done.

uristdwarf reviewed (changes requested):
First review

I've tagged some comments as [IMPORTANT] that must be changed before merging to dev, everything else is a suggestion.

rolysr commented on utils/metadata_notifier.cc:
```diff
@@ -0,0 +1,209 @@
+				fprintf(stderr, "Unknown packet type: %u\n", packetType);
+			}
+
+			recvBuffer.erase(recvBuffer.begin(), recvBuffer.begin() + 8 + dataLen);
```
This is a utility binary only for testing purposes, it is not intended to be used as official binary from our releases.

rolysr commented on utils/metadata_notifier.cc:
```diff
@@ -0,0 +1,209 @@
+	put16bit(&ptr, kCfgDefaultMasterTimeout);
+
+	return outpacket;
+}
```
Considered, but this is a util binary only for testing purposes.

rolysr commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	pt[3] = ip & 0xFF;
+	return std::to_string(pt[0]) + "." + std::to_string(pt[1]) + "." + std::to_string(pt[2]) + "." +
+	       std::to_string(pt[3]);
+}
```
Done.

rolysr commented on src/master/filesystem_operations.cc:
```diff
@@ -395,6 +406,31 @@ uint8_t fs_apply_access(uint32_t ts, inode_t inode) {
+
+	fullPath = "/" + fullPath;
+	return fullPath;
+}
```
Considered.

rolysr commented on src/cgi/sfs.cgi.in:
```diff
@@ -1739,6 +1742,86 @@ if "CS" in sectionset:
+                    servers.sort()
+
+                    if MBrev:
+                        servers.reverse()
```
Done.

rolysr commented on src/cgi/sfs.cgi.in:
```diff
@@ -1739,6 +1742,86 @@ if "CS" in sectionset:
+                for idx, (sf, hosts, ip_str, version_str) in enumerate(servers, start=1):
+                    out.append(f"""<tr>
+                        <td align="right">{idx}</td>
+                        <td align="left">{host}</td>
```
Done.

rolysr commented on src/cgi/sfs.cgi.in:
```diff
@@ -1739,6 +1742,86 @@ if "CS" in sectionset:
+
+            s = socket.socket()
+            s.connect((masterhost, masterport))
+            mysend(s, struct.pack(">LLL", CLTOMA_INOTIFIER_LIST, 4,0))
```
Done.

rolysr commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+			eptr->inputpacket.buffer.resize(kPacketHeaderSize);
+			eptr->inputpacket.startOffset = 0;
+			eptr->inputpacket.bytesLeft = kPacketHeaderSize;
+			matontserv_gotpacket(eptr, type, eptr->inputpacket.buffer.data(), size);
```
Done.

rolysr commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+		safs_pretty_syslog(LOG_NOTICE, "NTTOMA_GET_PATH_TYPE_INODE - inode %lu not found",
+		                   responseInode);
+		// Send back an empty path
+		responseData = matontserv_addpacket(eptr, MATONT_GET_PATH_TYPE_INODE, sizeof(uint64_t));
```
Done.

rolysr commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+	pt[3] = ip & 0xFF;
+	return std::to_string(pt[0]) + "." + std::to_string(pt[1]) + "." + std::to_string(pt[2]) + "." +
+	       std::to_string(pt[3]);
+}
```
Done.

rolysr commented on src/master/filesystem_operations.cc:
```diff
@@ -26,6 +26,7 @@
 #include <cstdarg>
 #include <cstddef>
 #include <cstdint>
+#include <filesystem>
```
Done.

rolysr commented on src/master/filesystem_node.cc:
```diff
@@ -26,6 +26,7 @@
 #include <cassert>
 #include <cstdint>
 #include <cstdlib>
+#include <filesystem>
```
Done.

rolysr commented on src/cgi/sfs.cgi.in:
```diff
@@ -1739,6 +1742,86 @@ if "CS" in sectionset:
+                    servers.sort()
+
+                    if MBrev:
+                        servers.reverse()
```
Done.

rolysr commented on src/cgi/sfs.cgi.in:
```diff
@@ -1739,6 +1742,86 @@ if "CS" in sectionset:
+
+                    try:
+                        host = (socket.gethostbyaddr(ip_str))[0]
+                    except Exception:
```
For this specific case, it is known that the exception could only come from the one command inside the `try`, then we could leave it like that.

rolysr commented on utils/metadata_notifier.cc:
```diff
@@ -0,0 +1,209 @@
+		free(pack->packet);
+		free(pack);
+	}
+}
```
Done.

rolysr commented on src/cgi/sfs.cgi.in:
```diff
@@ -1739,6 +1742,86 @@ if "CS" in sectionset:
+                        <td align="left">{host}</td>
+                        <td align="center">{ip_str}</td>
+                        <td align="center">{version_str}</td>
+                    </tr>""")
```
Done.

rolysr commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,524 @@
+		responseData = matontserv_addpacket(eptr, MATONT_GET_PATH_TYPE_INODE, sizeof(uint64_t));
+		put64bit(&responseData, responseInode);
+		put8bit(&responseData, FSNodeType::kUnknown);
+		memcpy(responseData, "", 1);
```
Done.

uristdwarf commented on src/master/filesystem_operations.cc:
```diff
@@ -106,6 +107,15 @@ void FilesystemOperationsBase::fs_changelog(uint32_t ts, const char *format, ...
+#ifdef METARESTORE
+	return false;
+#else
+	return matontserv_canexit();
```
Dead code

rolysr replied:
Done.

uristdwarf commented on src/data/sfsmaster.cfg.in:
```diff
@@ -322,6 +322,10 @@
 
 ##### EXPERIMENTAL:
 
+## Port to listen on for notifier connections.
+## (Default: 9418)
+# MATONT_LISTEN_PORT = 9418
+
```
[IMPORTANT] By default, this service should be disabled, rather than run automatically. You could set the default to 0 and check that on whether to enable the service or not

rolysr replied:
Done.

uristdwarf commented on src/protocol/SFSCommunication.h:
```diff
@@ -355,6 +355,28 @@ enum class SugidClearMode : uint8_t {
+// 0x00CC
+#define MATONT_GET_PATH_TYPE_INODE (PROTO_BASE+204)
+/// inode:64 nodetype:8 path:STDSTRING
+
```
Why not use the protocol with the versioning and macros to help build them?

[IMPORTANT] Also mark these as experimental as well.

rolysr replied:
This was for simplicity for the first iteration for also the development of the inotify application. For now I marked them as EXPERIMENTAL

uristdwarf commented on src/protocol/SFSCommunication.h:
```diff
@@ -1233,6 +1255,13 @@ enum class SugidClearMode : uint8_t {
+// 0x005F5
+#define MATOCL_INOTIFIER_LIST (1000U + 525U)
+// N * [ip:32 version:32]
+
```
[IMPORTANT] Also mark as experimental

rolysr replied:
Done.

uristdwarf commented on src/admin/list_inotifiers_command.cc:
```diff
@@ -0,0 +1,95 @@
+
+void ListInotifiersCommand::usage() const {
+	std::cerr << name() << " <master ip> <master port>" << std::endl;
+	std::cerr << "    Prints status of active inotify notifiers" << std::endl;
```
[IMPORTANT] Make sure to describe the command as EXPERIMENTAL

rolysr replied:
Done.

uristdwarf reviewed (changes requested):
Still some things should be changed.

lgsilva3087 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,507 @@
+		return;
+	}
+
+	// Convert the received 64-bit inode value to the runtime-configured inode_t type (may be
```
The inode_t is configured at build time.

rolysr replied:
Done.

lgsilva3087 commented on utils/metadata_notifier.cc:
```diff
@@ -0,0 +1,214 @@
+	return outpacket;
+}
+
+packetstruct *createGetPathTypeInodePacket(uint64_t inode) {
```
Should be inode_t.

rolysr replied:
Considered. But in the case of the expected behavior of the notifier application, it will always send 64-bits size integers referring to inodes, then it is master server responsibility to convert it to the current inode_t that it supports, the idea is to keep this util tool like this in order to better test and represent expected behavior, as for now the notifier application won't be compiled alongside with our saunafs project.

lgsilva3087 commented on utils/metadata_notifier.cc:
```diff
@@ -0,0 +1,214 @@
+	outpacket->bytesleft = psize;
+	uint8_t *ptr = outpacket->packet;
+	put32bit(&ptr, NTTOMA_GET_PATH_TYPE_INODE);
+	put32bit(&ptr, 8);
```
Preffer putInode which will select the correct inode_t type and size.

rolysr replied:
Considered. Please refer to my previous message.

lgsilva3087 commented on src/cgi/sfs.cgi.in:
```diff
@@ -32,6 +32,8 @@ CLTOMA_CSSERV_REMOVESERV = (PROTO_BASE + 524)
 CLTOCS_HDD_LIST_V2 = (PROTO_BASE + 600)
 CSTOCL_HDD_LIST_V2 = (PROTO_BASE + 601)
+CLTOMA_INOTIFIER_LIST = 1524
+MATOCL_INOTIFIER_LIST = 1525
```
I see you are using the modern way to serialize/deserialize the packets. In this case, the packets names/aliases should start with SAU_.

rolysr replied:
Done.

lgsilva3087 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,507 @@
+		return;
+	}
+
+	if (length < sizeof(uint64_t)) {
```
sizeof(inode_t)

rolysr replied:
Considered. The idea here is to keep notifier applications to send the inode on a 64-bit value, even if the master was not compiled to support 64-bits inodes. The idea was to make the notifier applications agnotic from the inode_t value that is supported on the master, which is responsible of truncating the inode value to the one that it is supporting. What can be done instead, is to add a check that if the master support inode_t = uint32_t, and it receives a value from a notifier that does not fit on 32-bits, then on that case consider that not inode found, as it is not supporting that size of inode, so it won't exist.

lgsilva3087 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,507 @@
+	// Convert the received 64-bit inode value to the runtime-configured inode_t type (may be
+	// uint32_t or uint64_t)
+	uint8_t *responseData;
+	uint64_t responseInode = get64bit(&data);
```
inode_t responseInode = getInode...

rolysr replied:
I could not be received for now as less that 64-bits, for the notifier application requirements.

lgsilva3087 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,507 @@
+
+	if (tcpstrlisten(listenSocket, ListenHost.c_str(), ListenPort.c_str(), 100) < 0) {
+		safs::log_err("master <-> notifiers module: can't listen on {}:{}",
+		              ListenHost.c_str(), ListenPort.c_str());
```
Consider not using .c_str() if you are using the modern logging functions.

rolysr replied:
Done.

lgsilva3087 commented on src/protocol/SFSCommunication.h:
```diff
@@ -1233,6 +1255,13 @@ enum class SugidClearMode : uint8_t {
 #define SAU_MATOCL_METADATASERVERS_LIST (1000U + 523U)
 // masterversion:32 data:(N * [ ip:32 hostname:STDSTRING version:32])
 
+// 0x005F4 [EXPERIMENTAL]
+#define CLTOMA_INOTIFIER_LIST (1000U + 524U)
+
+// 0x005F5 [EXPERIMENTAL]
+#define MATOCL_INOTIFIER_LIST (1000U + 525U)
+// N * [ip:32 version:32]
+
```
Consider SAU_ as prefix if they are using the modern way.

rolysr replied:
Done.

lgsilva3087 commented on utils/metadata_notifier.cc:
```diff
@@ -0,0 +1,214 @@
+	uint8_t *startptr;
+	uint32_t bytesleft;
+	uint8_t *packet;
+};
```
Please avoid the C-like linked lists in new code. They just make the things harder to maintain.

rolysr replied:
Done.

lgsilva3087 reviewed (commented):
Please check my comments.

rolysr committed:
refactor(utils): Modernize metadata notifier

This commit modernizes the metadata notifier util binary in order
to improve its maintainability and support as adding more memory
safety on previous dangerous code sections requiring manual memory
management.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

dmga44 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,512 @@
+	uint32_t lastRead, lastWrite;
+	std::array<uint8_t, kPacketHeaderSize> headerBuff{};
+	PacketStruct inputPacket;
+	std::queue<PacketStruct> outputPackets;
```
Consider using InputPacket and OutputPacket classes (like in matoclserv.cc):
```
	InputPacket inputPacket{MaxPacketSize};  ///< InputPacket for reading data from the client
	std::list<OutputPacket> outputPackets;  ///< List of output packets
```

You can get rid of the PacketStruct struct and simplify the NotifierConnectionMode enum.

rolysr replied:
Done.

dmga44 reviewed (approved):
Great job, guys.

Please check my suggestion.

rolysr committed:
chore: Remove const qualifier from InputPacket

This commit removes the uneeded const qualifier from the maxPacketSize_
member in the InputPacket struct as its immutability is still enforced
by making maxPacketSize_ private and only settable via the constructor.
This change also allows to better integrate InputPacket on master
communication related entries for other SaunaFS components, allowing
compiling these struct entries with STL containers such as std::deque

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

lgsilva3087 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,478 @@
+#include "common/sockets.h"
+#include "config/cfg.h"
+#include "master/filesystem_node.h"
+#include "master/filesystem_operations.h"
```
Please include:
```cpp
#include "master/filesystem_operations_interface.h"
```
It is lightweight and provides the needed functions.

rolysr replied:
Done.

rolysr replied:
Done.

lgsilva3087 commented on src/master/matontserv.h:
```diff
@@ -0,0 +1,33 @@
+#include <cstdint>
+
+#include "common/inotifier_list_entry.h"
+#include "common/metadataserver_list_entry.h"
```
Looks unused in clangd.

rolysr replied:
Done.

lgsilva3087 commented on src/master/matontserv.cc:
```diff
@@ -0,0 +1,478 @@
+	uint16_t timeout;
+	std::string serviceStrIp;
+	uint32_t version;
+	uint32_t serviceIp;
+	std::string config;
+
+	MatontservEntry(int sock_, uint32_t now)
+	    : mode(NotifierConnectionMode::HEADER),
+	      sock(sock_),
+	      pollFdIndex(-1),
+	      lastRead(now),
+	      lastWrite(now),
+	      headerBuff{0},
+	      inputPacket(kMaxPacketSize),
+	      timeout(10),
+	      serviceStrIp(),
+	      version(0),
+	      serviceIp(0),
+	      config() {}
+};
```
Consider this to remove most of the warnings:

```cpp
struct MatontservEntry {
	NotifierConnectionMode mode = NotifierConnectionMode::HEADER;
	int sock;
	int32_t pollFdIndex{-1};
	uint32_t lastRead{0};
	uint32_t lastWrite{0};
	std::array<uint8_t, kPacketHeaderSize> headerBuff{0};
	InputPacket inputPacket{kMaxPacketSize};
	std::queue<OutputPacket> outputPackets;
	uint16_t timeout{10};
	std::string serviceStrIp;
	uint32_t version{};
	uint32_t serviceIp{};
	std::string config;

	MatontservEntry(int sock_, uint32_t now) : sock(sock_), lastRead(now), lastWrite(now) {}
};
```

rolysr replied:
Done.

lgsilva3087 commented on src/master/matoclserv.cc:
```diff
@@ -1028,6 +1029,13 @@ void matoclserv_mlog_list(matoclserventry *eptr, const uint8_t *data, uint32_t l
+	(void)data;
+	(void)length;
+	cltoma::inotifierList::deserialize(data, length);
+	matoclserv_createpacket(eptr, matocl::inotifierList::build(matontserv_inotifiers()));
```
Nice.

lgsilva3087 commented on src/master/matoclserv.cc:
```diff
@@ -1028,6 +1029,13 @@ void matoclserv_mlog_list(matoclserventry *eptr, const uint8_t *data, uint32_t l
 
+void matoclserv_inotifier_list(matoclserventry *eptr, const uint8_t *data, uint32_t length) {
+	(void)data;
+	(void)length;
```
The are used below, maybe the cast to void is not needed anymore.

rolysr replied:
Done.

lgsilva3087 reviewed (approved):
It is looking good. Please see my minor comments.

antuan96314 commented on src/common/input_packet.h:
```diff
@@ -105,7 +105,7 @@ struct InputPacket {
 private:
 	/// Maximum accepted length of a message.
-	const uint32_t maxPacketSize_;
+	uint32_t maxPacketSize_;
```
@rolysr  Why is this needed?

rolysr replied:
This change was necessary to resolve a compilation issue that occurred when adding `InputPacket` as a member of `MatontservEntry` and using it with `std::queue` (or `std::deque`). If a struct contains a member with a non-static const field, the compiler deletes the copy assignment operator for that struct. As a result, standard containers like std::queue and `std::deque`—which require their elements to be copy-assignable or move-assignable—cannot be used with such structs. By removing the const qualifier from the private field in `InputPacket`, the compiler is able to generate the necessary move and copy assignment operators, making the struct compatible with these standard containers and ensuring correct behavior. Also, after taking a look to that field on the InputPacket class, it was noticed that it was not needed to make it constant, as it is a private field with now public or exposed getter, its value is only set on the constructor and it was not set or modified in any other part of the current code. The other reason to force this change was to use `std::queue` as container for `OutputPackets` instead of `std::list` which is less efficient in the needed use case.
